packages/runtime/container-runtime/src/compatUtils.ts

```diff
@@ -0,0 +1,194 @@
+ * by **default** will be in effect starting with 3.0. Importantly though, N/N-1 compatibility is still guaranteed with the
+ * proper configurations set.
+ *
+ * TODO: This should be updated to "2.0.0" when 3.0 is released.
+ */
+export const defaultCompatibilityMode = "pre-3.0-default";
+
+/**
+ * The current default set of configurations if no compatibility mode is provided.
+ * Since both compatibilityMode and the cross-client compat policy was introduced during 2.x's lifespan, we will maintain the
+ * set of default configurations that were in place before compatibilityMode and the policy were introduced.
+ * TODO: This can be removed after 3.0 is released.
+ */
+const defaultConfigsForPreFF3: IContainerRuntimeOptionsVersionDependent = {
+	gcOptions: {},
+	flushMode: FlushMode.TurnBased,
+	compressionOptions: enabledCompressionConfig,
+	enableRuntimeIdCompressor: undefined as unknown as "on" | "delayed",
+	enableGroupedBatching: true,
+	explicitSchemaControl: false,
+};
+
+/**
+ * Subset of the IContainerRuntimeOptionsInternal properties which are version-dependent.
+ *
+ * @remarks
+ * When a new option is added to IContainerRuntimeOptionsInternal, we must consider if it's a version-dependent option.
+ * If it's considered version-dependent, then a corresponding entry must be added to `versionDependentOptionConfigs`. If not, then
+ * it must be omitted from this type.
+ *
+ * Note: We use `Omit` instead of `Pick` to ensure that all new options are included in this type by default. If any new properties
+ * are added to IContainerRuntimeOptionsInternal, they will be included in this type unless explicitly omitted. This will prevent
+ * us from forgetting to account for any new properties in the future.
+ */
+export type IContainerRuntimeOptionsVersionDependent = Required<
+	Omit<
+		IContainerRuntimeOptionsInternal,
+		| "chunkSizeInBytes"
+		| "maxBatchSizeInBytes"
+		| "loadSequenceNumberVerification"
+		| "summaryOptions"
+		| "compatibilityMode"
+	>
+>;
+
+/**
+ * Default configurations for version-dependent options based on the compatibilityMode.
+ */
+interface IVersionDependentOptionConfig<
+	K extends keyof IContainerRuntimeOptionsVersionDependent,
+> {
+	/**
+	 * The minimum version of the container runtime that is required to use the modern config for the
+	 * option. This will compared with the compatibilityMode to determine if the modern or legacy config
+	 * should be used by default.
+	 * This should be undefined if no versions are ready to use the modern config by default. In this case,
+	 * we will always default to the legacy config.
+	 */
+	minVersionForModernConfig: string | undefined;
+	/**
+	 * The default config for the version-dependent option that ensures clients that do not understand the
+	 * modern config will not break.
+	 * This is the config that will be used by default when the compatibilityMode is less than minVersionForModernConfig.
+	 */
+	legacyConfig: IContainerRuntimeOptionsVersionDependent[K];
```

Comment:
i don't really like the use of the word legacy here as legacy already has a very specific meaning in our code base, and it is different for the meaning here

Reply:
i think my below comment fixes this: https://github.com/microsoft/FluidFramework/pull/24248/files#r2042841307